softicar/platform
# PLAT-606 Eliminate dependency on commons-fileupload

### Summary

1. File uploads are now implemented with JDK (`javax.servlet`) classes, instead of `commons-fileupload`.
   - The dependency on `commons-fileupload` was removed.
2. Unified Jetty server creation for (1) unit tests and (2) development/prototyping.
   - The existing file upload unit tests are now representative for all of our Jetty use cases.
   - `AjaxTestingServer` now extends `AbstractStandAloneServletServer`.
   - Introduced `StandAloneServletServerConfiguration`, and migrated default value definitions from `AbstractStandAloneServletServer`.
   - Removed some related warning suppressions and TODOs that became unnecessary or fixed.
3. `AbstractStandAloneServletServer` now uses a random port by default.
   - No longer crashes because the former default port of `0` was invalid.
4. Eliminated obsolete/orphaned `IAjaxTestingServerEnvironment`, for the sake of simplicity.

### Notes to the reviewer
- We now use built-in JDK classes which require us to comply with Servlet API `3.x`.
- This entails the specification of a "Multipart Configuration" to handle requests, which was not necessary before.
- We realize this via different approaches, depending on the employed web server / servlet container:
  - For Jetty, we set a default configuration in `AbstractStandAloneServletServer`.
  - For Tomcat, we use `@MultipartConfig` on `WebServiceServlet`.
  - A singular solution for both (all) cases could _not_ be found.
- No file upload unit test was added in this PR.
  - For such a test to be practically relevant, we'd need a Tomcat container during unit test execution.
  - This would be quite some effort, so manual testing was preferred, for now.

### Test Plan ### 

**1. Test on Jetty**
- Run this:
  ```
  public class PlatformTestServlet extends HotDeploymentWebServiceServlet {

    public PlatformTestServlet() {

      database.applyFixture(this::apply);
    }

    private PlatformTestServlet apply() {

      TestFixtureRegistry registry = new TestFixtureRegistry(new CoreModuleTestFixture());
      registry.registerIfMissing(DemoModuleTestFixture::new);
      return this;
    }

    public static void main(String[] args) {

      new HotDeploymentWebServiceServer(PlatformTestServlet.class)//
        .setRequestString("service?id=" + EmfSourceCodeReferencePoints.getUuidOrThrow(PageService.class))
        .setPort(9000)
        .startAndJoin();
    }
  }
  ```
- Log in with `admin.user` / `test`
- Head to `System / Demo / Demo Invoices`
- Edit an arbitrary row
- Add a single `Document`, and several `Attachments`, and hit `Save`
- Make sure that the files are displayed in the Management Div, and can be downloaded.

**2. Test on Tomcat**
- Preparation
  - Install Tomcat 9.0.34, with the usual configuration, and start it: `~/apache-tomcat/bin/catalina.sh start`
  - Start a MariaDB docker container: `docker run --name mariadb-test -e MYSQL_ROOT_PASSWORD=iamroot -d mariadb:10.5.8`
    - Find the IP of that container; below, `172.20.0.2` is assumed.
  - Save this as `.softicar/config.properties`:
    ```
    com.softicar.db.server.address = 172.20.0.2
    com.softicar.db.username = root
    com.softicar.db.password = iamroot
    com.softicar.common.mail.sending.enabled = false
    com.softicar.common.mail.dumping.enabled = true
    ```
- Publish the platform locally:
  ```
  cd ~/workspace/platform
  ./gradlew clean assemble
  ./gradlew publish -Ppublish.target=local -Pversion=13.0.0-SNAPSHOT
  ```
- In a checkout of `com.softicar.eas`:
  - Change `build.gradle` to use the local snapshot release of the platform.
  - Change `com.softicar.eas.root/build.gradle` by adding this line: `api "com.softicar.platform:platform-demo-module:$versions.platform"`
- Run:
  ```
  cd ~/workspace/com.softicar.eas
  ./gradlew clean war
  cp ~/workspace/com.softicar.eas/build/libs/com.softicar.eas-??.??.??.?.war ~/apache-tomcat/webapps/portal.war
  ```
- Create this class in `com.softicar.eas` (or a derived local project), replace `172.20.0.2` with the IP of the MariaDB container, and run it:
  ```
  public class MariaDbDatabaseSetupExecutor {

    public static void main(String[] args) {

      var executor = new MariaDbDatabaseSetupExecutor();
      executor.executeDatabaseStructureSetup();
      executor.executeDatabaseContentSetup();
      Log.finfo("Database setup complete.");
    }

    private DbDatabase database;

    public MariaDbDatabaseSetupExecutor() {

      this.database = new DbDatabaseBuilder().setServerType(DbServerType.MARIA_DB).setHostname("172.20.0.2")
          .setUsername("root").setPassword("iamroot").build();
    }

    public void executeDatabaseStructureSetup() {

      new MariaDbDatabaseSetup(database).execute();
    }

    public void executeDatabaseContentSetup() {

      try (DbDatabaseScope databaseScope = new DbDatabaseScope(database)) {
        try (DbConnectionOverrideScope connectionScope = new DbConnectionOverrideScope(database)) {
          TestFixtureRegistry registry = new TestFixtureRegistry(new CoreModuleTestFixture());
          registry.registerIfMissing(DemoModuleTestFixture::new);
        }
      }
    }
  }
  ```
- Enter: `2`, `y`, `1`, `4`
- Make sure that no errors are printed.
- Log in at `http://localhost:8080/portal/service?id=95cf1a1b-c12e-4594-9d20-783988fe32b9&moduleInstance=1&page=23254f8d-92ea-4895-8aef-2837a8b3f368`
- Perform the same file uploading test as above.
